src/jit/lclvars.cpp

```diff
@@ -2297,6 +2298,115 @@ void Compiler::lvaSetStruct(unsigned varNum, CORINFO_CLASS_HANDLE typeHnd, bool
     }
 }
 
+//------------------------------------------------------------------------
+// lvaSetClass: set or update class information for a local var.
+//
+// Arguments:
+//    varNum -- number of the variable
+//    clsHnd -- class handle to use in set or update
+//    isExact -- true if class is known exactly
+//
+// Notes:
+//
+//    This method can be viewed as a rudimentary lattice meet in a
+//    type lattice.
+//
+//    Updates currently should only happen for single-def user args or
+//    locals, when we are processing the expression actually being
+//    used to initialize the local (or inlined arg). The update will
+//    change the local from the declared type to the type of the
+//    initial value.
+//
+//    These updates should always *improve* what we know about the
+//    type, that is making an inexact type exact, or changing a type
+//    to some subtype. However the jit lacks precise type information
+//    for shared code, so ensuring this is so is currently not
+//    possible.
+
+void Compiler::lvaSetClass(unsigned varNum, CORINFO_CLASS_HANDLE clsHnd, bool isExact)
+{
+    noway_assert(varNum < lvaCount);
+    assert(clsHnd != nullptr);
+
+    LclVarDsc* varDsc = &lvaTable[varNum];
+    assert(varDsc->lvType == TYP_REF);
+
+    // If previous type was exact, nothing to update.  Would like
+    // to verify new type is compatible but can't do this yet.
+    if (varDsc->lvClassIsExact)
+    {
+        assert(varDsc->lvClassHnd != nullptr);
+        return;
+    }
+
+    // If we didn't have a type yet, use the one we have now.
+    if (varDsc->lvClassHnd == nullptr)
+    {
+        assert(!varDsc->lvClassIsExact);
+        JITDUMP("\nlvaSetClass: setting class for V%02i to (%p) %s %s\n", varNum, clsHnd,
+                info.compCompHnd->getClassName(clsHnd), isExact ? " [exact]" : "");
+
+        varDsc->lvClassHnd     = clsHnd;
+        varDsc->lvClassIsExact = isExact;
+        return;
+    }
+
+    // Are we updating the type?
+    if (varDsc->lvClassHnd != clsHnd)
```

Comment:
> updates can happen for single-def locals too, as we move from the signature type to the type of the initializing value

So generally you'd expect zero or one calls to the overload that doesn't take a tree, followed by zero or one calls to the overload that does take a tree?